### PR TITLE
ocamlPackages : update some packages

### PR DESCRIPTION
**Pull request by thufschmitt**
###### Things done
- [X] Tested using sandboxing
  ([nix.useChroot](http://nixos.org/nixos/manual/options.html#opt-nix.useChroot) on NixOS,
    or option `build-use-chroot` in [`nix.conf`](http://nixos.org/nix/manual/#sec-conf-file)
    on non-NixOS)
- Built on platform(s)
  - [X] NixOS
  - [ ] OS X
  - [ ] Linux
- [X] Tested compilation of all pkgs that depend on this change using `nix-shell -p nox --run "nox-review wip"`
- [X] Tested execution of all binary files (usually in `./result/bin/`)
- [X] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

---

### CODE REVIEW

**Comment by LnL7** [comment]
It looks like most of the configuration for the compiler versions is the same, this should be generalized.

**Comment by regnat** [comment]
Indeed, I was to eager to submit it upstream, I forgot that I did a dirty copy/paste.

**Comment by ocamlpro-Hufschmitt** [comment]
I rebased onto master to fix a conflict.

Is this PR ok to be merged ?

File: pkgs/development/compilers/ocaml/generic.nix [hunk 1]
```diff
@@ -0,0 +1,81 @@
...
+
+stdenv.mkDerivation (args // rec {
+
+  x11env = buildEnv { name = "x11env"; paths = [libX11 xproto]; };
+  x11lib = x11env + "/lib";
+  x11inc = x11env + "/include";
+
+  inherit name;
+
+  src = fetchurl {
```
Comment (LnL7) [inline pkgs/development/compilers/ocaml/generic.nix]: for generic things like this it's useful to allow both specifying a `src` or `version` that then uses the default source.

Comment (regnat) [inline pkgs/development/compilers/ocaml/generic.nix]: Ok.

I just splitted the version into three arguments (major, minor and patch) as it needs to be splitten to build the download url.

**Comment by joachifm** [comment]
Travis error looks related:

> while evaluating the file ‘/home/travis/.nox/nixpkgs/pkgs/development/compilers/ocaml/4.01.0.nix’:
anonymous function at /home/travis/.nox/nixpkgs/pkgs/development/compilers/ocaml/generic.nix:1:1 called without required argument ‘minor_version’, at /home/travis/.nox/nixpkgs/pkgs/development/compilers/ocaml/4.01.0.nix:1:1

**Comment by regnat** [comment]
Oh indeed, I had forgotten to update the 4.01 version.
It should be OK now

**Pull request merged**